### PR TITLE
run: Show help even if no oci image was given

### LINKED ISSUE #2005: containerized-gadgets: `run -h` doesn't work if gadget image isn't provided
The help of the `run` command isn't printed the gadget image is not provided:

```bash
$ go run -exec "sudo -E" ./cmd/ig/. run -h
INFO[0000] Experimental features enabled                
Error: calling custom parser: no definition provided
exit status 1
```

It works fine when both are provided:

```bash
$ go run -exec "sudo -E" ./cmd/ig/. run ghcr.io/inspektor-gadget/gadget/trace_exec:latest  -h
INFO[0000] Experimental features enabled                
Run a containerized gadget

Usage:
  ig run [flags]

Flags:
      --authfile string         (default "/var/lib/ig/config.json")
  -c, --containername string   Show only data from containers with that name (default "")
  -F, --filter strings         Filter rules
                                 A filter can match any column using the following syntax
                                   columnName:value       - matches, if the content of columnName equals exactly value
                                   columnName:!value      - matches, if the content of columnName does not equal exactly value
                                   columnName:>=value     - matches, if the content of columnName is greater than or equal to the value
                                   columnName:>value      - matches, if the content of columnName is greater than the value
                                   columnName:<=value     - matches, if the content of columnName is less than or equal to the value
                                   columnName:<value      - matches, if the content of columnName is less than the value
                                   columnName:~value      - matches, if the content of columnName matches the regular expression 'value'
                                                            see [https://github.com/google/re2/wiki/Syntax] for more information on the syntax
                               
  -h, --help                   help for run
      --host                   Show data from both the host and containers
  -o, --output string          Output format (json, jsonpretty, yaml, columns).
                               
                               Columns (columns)
                                 The output of the gadget is formatted in human readable columns.
                                 You can optionally specify the columns to output using '-o columns=col1,col2,col3' etc.
                                 Columns can be prefixed with '+' or '-' to add or remove columns relative to the default columns.
                                 Columns 'k8s' and 'runtime' are expanded to all available columns for the respective environment.
                                   Available columns:
                                     comm
                                     gid
                                     k8s.container
                                     k8s.hostnetwork
                                     k8s.namespace
                                     k8s.node
                                     k8s.pod
                                     mntns
                                     pid
                                     ppid
                                     runtime.containerId
                                     runtime.containerImageName
                                     runtime.containerName
                                     runtime.runtimeName
                                     timestamp
                                     uid
                                   Default columns: runtime.containerName,pid,ppid,comm,uid,gid
                               
                               JSON (json)
                                 The output of the gadget is returned as raw JSON
                               JSON Prettified (jsonpretty)
                                 The output of the gadget is returned as prettified JSON
                               YAML (yaml)
                                 The output of the gadget is returned as YAML
                               
                                (default "columns")
  -t, --timeout int            Number of seconds that the gadget will run for, 0 to disable

Global Flags:
      --auto-mount-filesystems          Automatically mount bpffs, debugfs and tracefs if they are not already mounted
      --auto-sd-unit-restart            Automatically run in a privileged systemd unit if lacking enough capabilities
      --auto-wsl-workaround             Automatically find the host procfs when running in WSL2
      --containerd-namespace string     Containerd namespace to use (default "k8s.io")
      --containerd-socketpath string    Containerd CRI Unix socket path (default "/run/containerd/containerd.sock")
      --crio-socketpath string          CRI-O CRI Unix socket path (default "/run/crio/crio.sock")
      --docker-socketpath string        Docker Engine API Unix socket path (default "/run/docker.sock")
      --metrics-listen-address string   Address to serve prometheus metrics on (default "0.0.0.0:2223")
      --metrics-path string             Path to export prometheus metrics on (default "/metrics")
      --podman-socketpath string        Podman Unix socket path (default "/run/podman/podman.sock")
  -r, --runtimes string                 Container runtimes to be used separated by comma. Supported values are: docker, containerd, cri-o, podman (default "docker,containerd,cri-o,podman")
  -v, --verbose                         Print debug information
```

This happens because we have some custom logic to print the help after the parser is created, this in order to provide information about the columns that are supported by the gadget

https://github.com/inspektor-gadget/inspektor-gadget/blob/f832e161176f8cc85f5e92f7c6ee25cbd5803b7c/cmd/common/registry.go#L260

## Solution

We should print some the help (without the columns) when those parameters aren't specified.  Something like the following works, but some additional logic is needed to indicate the available columns are only printed if `--prog` and `--definition` are passed.

```diff
$ git diff
diff --git cmd/common/registry.go cmd/common/registry.go
index 5a5239c3..5056f995 100644
--- cmd/common/registry.go
+++ cmd/common/registry.go
@@ -241,6 +241,9 @@ func buildCommandFromGadget(
                                var err error
                                parser, err = c.CustomParser(gadgetParams, cmd.Flags().Args())
                                if err != nil {
+                                       if showHelp, _ := cmd.Flags().GetBool("help"); showHelp {
+                                               return cmd.Help()
+                                       }
                                        return fmt.Errorf("calling custom parser: %w", err)
                                }
                        }
```

### PR DESCRIPTION
**Pull request by burak-ok**
# run: Show help even if no oci image was given

resolves #2005

When we don't specify an oci image the `run` gadget still tried to show all available columns. For this it has to look into the BTF code stored in the oci image.

To resolve the problem this PR implements the following behavior:
1. If no oci image is given the help is printed without any columns. Additionally the message `To see available columns for a specific gadget image, the oci image needs to specified as an argument` will be shown in the help output (See Testing done)
2. If a valid oci image is given, the help behaves like always did
3. If an invalid oci image is given, the help prints an error if the oci image url is malformed, can't be found, ....

### Without oci image
```bash
➭ go run --exec "sudo -E" ./cmd/ig/... run --help
INFO[0000] Experimental features enabled                
Run a containerized gadget
  To see available columns for a specific gadget image, the oci image needs to specified as an argument

Usage:
  ig run [flags]

Flags:
      --authfile string         (default "/var/lib/ig/config.json")
//*****SNIP*****
  -o, --output string          Output format (json, jsonpretty, yaml).
                               
                               JSON (json)
                                 The output of the gadget is returned as raw JSON
                               JSON Prettified (jsonpretty)
                                 The output of the gadget is returned as prettified JSON
                               YAML (yaml)
                                 The output of the gadget is returned as YAML
                               
                                (default "columns")
  -t, --timeout int            Number of seconds that the gadget will run for, 0 to disable

```
### With valid oci image
```bash
➭ go run --exec "sudo -E" ./cmd/ig/... run ghcr.io/inspektor-gadget/gadget/trace_open:latest --help
INFO[0000] Experimental features enabled                
Run a containerized gadget

Usage:
  ig run [flags]

Flags:
      --authfile string         (default "/var/lib/ig/config.json")
//*****SNIP*****
  -o, --output string          Output format (jsonpretty, yaml, json, columns).
                               
                               Columns (columns)
                                 The output of the gadget is formatted in human readable columns.
                                 You can optionally specify the columns to output using '-o columns=col1,col2,col3' etc.
                                 Columns can be prefixed with '+' or '-' to add or remove columns relative to the default columns.
                                 Columns 'k8s' and 'runtime' are expanded to all available columns for the respective environment.
                                   Available columns:
                                     comm
                                     flags
//*****SNIP*****
                                     runtime.containerImageName
                                     runtime.containerName
                                     runtime.runtimeName
                                     timestamp
                                     uid
                                   Default columns: runtime.containerName,pid,comm,uid,gid,ret,fname
                               
                               JSON (json)
                                 The output of the gadget is returned as raw JSON
                               JSON Prettified (jsonpretty)
                                 The output of the gadget is returned as prettified JSON
                               YAML (yaml)
                                 The output of the gadget is returned as YAML
                               
                                (default "columns")
```
### With invalid oci image url
```bash
➭ go run --exec "sudo -E" ./cmd/ig/... run ghcr.io/inspektor-ga --help                             
INFO[0000] Experimental features enabled                
Error: getting gadget info: get ebpf program and definition: get ebpf program: pull image: download to local repository: failed to resolve ghcr.io/inspektor-ga:latest: GET "https://ghcr.io/v2/inspektor-ga/manifests/latest": response status code 400: name invalid: invalid repository name
exit status 1
```

### CODE REVIEW

File: cmd/common/registry.go [hunk 1]
```diff
@@ -259,6 +259,14 @@ func buildCommandFromGadget(
 			// Before running the gadget, we need to get the gadget info to create the
 			// parser based on it
 			if isRunGadget {
+				if len(args) == 0 {
+					if showHelp, _ := cmd.Flags().GetBool("help"); showHelp {
+						additionalMessage := "To see available columns for a specific gadget image, the oci image needs to specified as an argument"
```
Comment (mauriciovasquezbernal) [inline cmd/common/registry.go:264]: I think we can make this message more generic. 

```suggestion
						additionalMessage := "Specify the gadget image to get more information about it"
```

File: cmd/common/registry.go [hunk 2]
```diff
@@ -259,6 +259,14 @@ func buildCommandFromGadget(
...
+						cmd.Long = fmt.Sprintf("%s\n  %s", cmd.Short, additionalMessage)
```
Comment (mauriciovasquezbernal) [inline cmd/common/registry.go:265]: I'd rather do:

```suggestion
						cmd.Long = fmt.Sprintf("%s\n\n%s", cmd.Short, additionalMessage)
```

**Review by mauriciovasquezbernal — APPROVED** [review]
Two minor comments, but LGTM already. Btw, I don't think columns is the only gadget-specific information we'll be showing, for instance I think we can also include:
- A custom gadget description 
- A custom gadget usage guide 
- Supported modes (trace, snapshot, top, etc)

And more. For sure that's work TBD, but this PR is a step in the direction.

Thanks for handling this!

**Comment by burak-ok** [comment]
> Btw, I don't think columns is the only gadget-specific information we'll be showing, for instance I think we can also include:
> 
> * A custom gadget description
> * A custom gadget usage guide
> * Supported modes (trace, snapshot, top, etc)
> 
> And more. For sure that's work TBD

Do we want to have an additional command for all of that?
Something like:
```bash
$ ig image info ghcr.io/inspektor-gadget/my-gadget:latest
Description: My newest gadget which can do everything
Author: Burak Ok
Columns: ...
Parameters: ...
Size: ...
```

**Pull request merged**